pkg/agent/openflow/network_policy.go

```diff
@@ -184,100 +184,202 @@ type conjMatchFlowContext struct {
 }
 
-// addAction adds the specified conjunction into conjunctiveMatchFlow's actions, and then updates the conjunctive
-// match flow on the switch. It also installs default drop flow if dropTable is not nil, and the dropFlow is not
-// installed before.
-func (ctx *conjMatchFlowContext) addAction(action *conjunctiveAction) error {
-	// Check if the conjunction exists in conjMatchFlowContext actions or not. If yes, return nil directly.
-	// Otherwise, add the new action, and update the Openflow entry.
+// addAction adds the specified policyRuleConjunction into conjunctiveMatchFlow's actions, and then returns the flowChange.
+func (ctx *conjMatchFlowContext) addAction(action *conjunctiveAction) *flowChange {
+	// Check if the conjunction exists in conjMatchFlowContext actions or not. If yes, return nil immediately.
 	_, found := ctx.actions[action.conjID]
 	if found {
 		return nil
 	}
 
-	// Install or update Openflow entry for the new conjunctiveAction.
-	actions := make([]*conjunctiveAction, 0, len(ctx.actions)+1)
+	// Append current conjunctive action to the existing actions, and then calculate the conjunctive match flow changes.
+	actions := []*conjunctiveAction{action}
 	for _, act := range ctx.actions {
 		actions = append(actions, act)
 	}
-	actions = append(actions, action)
-	err := ctx.installOrUpdateFlow(actions)
-	if err != nil {
-		return err
-	}
-	ctx.actions[action.conjID] = action
-
-	return nil
+	return ctx.createOrUpdateConjunctiveMatchFlow(actions)
 }
 
-func (ctx *conjMatchFlowContext) addDenyAllRule(ruleID uint32) error {
+func (ctx *conjMatchFlowContext) addDenyAllRule(ruleID uint32) {
 	if ctx.denyAllRules == nil {
 		ctx.denyAllRules = make(map[uint32]bool)
 	}
 	ctx.denyAllRules[ruleID] = true
-	return nil
+}
+
+func (ctx *conjMatchFlowContext) delDenyAllRule(ruleID uint32) {
+	// Delete the DENY-ALL rule if it is in context denyAllRules.
+	_, found := ctx.denyAllRules[ruleID]
+	if found {
+		delete(ctx.denyAllRules, ruleID)
+	}
+}
+
+type changeType int
+
+const (
+	add changeType = iota
+	modify
+	deletion
+)
+
+// flowChange maintains the expected Openflow entry and flow operation type that need to be applied on the OVS bridge.
+type flowChange struct {
+	flow       binding.Flow
+	changeType changeType
+}
+
+// actionChange describes the changed action of the conjunctive match flow, and the change type.
+type actionChange struct {
+	action     *conjunctiveAction
+	changeType changeType
+}
+
+type conjMatchFlowContextChange struct {
+	context *conjMatchFlowContext
+	// ctxType is the changed type of the conjMatchFlowContext. Its value includes "add", "modify" and "deletion". Add the
+	// context into the globalConjMatchFlowCache if the ctxType is "add", and remove from the globalConjMatchFlowCache
+	// if it is "deletion".
+	ctxType changeType
+	// matchFlow is the changed conjunctive match flow which needs to be realized on the OVS bridge. It is used to reset
+	// conjMatchFlowContext.flow.
+	matchFlow *flowChange
```

Comment:
I know it's a bit verbose, but please specify which member can be `nil` and which one cannot (and what it means to be `nil`).

Reply:
In particular, it seems that `actChange` is never `nil`

Reply:
Correct, actChange is never nil, for the action is actually representing the changed policyRuleConjunction, and it is what triggered the current change.  I have add comments to specific if it could be nil or not for the members,